README.md
# [Creators.TF](https://creators.tf) [Community Launcher](https://creators.tf/launcher) (Alpha)

[![](https://github.com/ampersoftware/Creators.TF-Community-Launcher/actions/workflows/ci.yml/badge.svg)](https://github.com/ampersoftware/Creators.TF-Community-Launcher/actions/workflows/ci.yml)
[![](https://github.com/ampersoftware/Creators.TF-Community-Launcher/actions/workflows/release.yml/badge.svg)](https://github.com/ampersoftware/Creators.TF-Community-Launcher/actions/workflows/release.yml)

## This repository contains the source code and releases for the [Creators.TF](https://creators.tf) [Community Launcher](https://creators.tf/launcher)

### Core features:
* Launcher auto-updating
* Custom mod support
* GitHub release mod support
* More details on our [website page](https://creators.tf/launcher)

### Useful links for this repository:
* [Releases](https://github.com/ampersoftware/Creators.TF-Community-Launcher/releases) -- [Latest release](https://github.com/ampersoftware/Creators.TF-Community-Launcher/releases/latest)
* [Changelogs](https://github.com/ampersoftware/Creators.TF-Community-Launcher/blob/master/changelog.md)
* [License](https://github.com/ampersoftware/Creators.TF-Community-Launcher/blob/master/LICENSE)
* [Code of Conduct](https://github.com/ampersoftware/Creators.TF-Community-Launcher/blob/master/CODE_OF_CONDUCT.md) (TL;DR; common sense)

### Maintainers
* [Jota](https://github.com/jota11) - jota@creators.tf - UI and UX Lead
* [rob5300](https://github.com/rob5300) - rob5300@creators.tf - Core Feature Lead

### Contributing
If you have something to contribute, please, open a pull request! Make sure your PR tries to solve/improve one thing at a time and isn't messy. Pull requests should always have a specific goal to accomplish.

Please do NOT make a pull request or change that only seeks to change code styling or do "housekeeping" as these will be denied and closed.

If you have a bug report or a suggestion, open an issue! Try to use the templates and give good information.
The launcher is currently in an alpha state so any help is appreciated.

## Launching and Building from Source
If you want to build the launcher yourself, you need to have [Node.js](https://nodejs.org/en/download/) installed, which should also include [npm](https://www.npmjs.com/get-npm).
Clone the repository using git *(or download from this page)* and do `npm install` in the main repository using your system's CLI to download the required packages.

You can start the launcher for testing purposes without building via `npm run compileAndStart`.

You can build for Windows or Linux via the included build scripts using `npm run buildWindows` or `npm run buildLinux`.
If you need, you can edit these in [package.json](https://github.com/ampersoftware/Creators.TF-Community-Launcher/blob/master/package.json#L9) to build to your liking, but changes made to these should not be committed if you plan to make a PR.


changelog.md
## Version 0.3.0 - (ALPHA)
### 08/10/2021 - 19:10 (GMT)
- Added new "Unique Rockets" mod (by AtomicTEM)
- Added new "Manned Machines Robot Model Fixes" mod (by HiGPS)
- Fixed GitHub collection mods not updating with all the correct assets/collection type.
- Disabled all animations to users with the "Reduced Motion" option enabled in their OS.
- Changed how GitHub mods get their version number. (This will trigger a fake update for users with a GitHub mod installed).
- Fixed mod's display version not being changed after an update.

## Version 0.2.9 - (ALPHA)
### 01/10/2021 - 21:19 (GMT)
- Fixes issue #71 , "Certificate has Expired" error when trying to download any creators.tf server content.

## Version 0.2.8 - (ALPHA)
### 08/09/2021 - 19:45 (GMT)
- Changed icons to use the new logo
- Fixed a visual bug on collection version mods *(currently only mastercomfig)* not showing the correct version installed
- GitHub mods now display their respective release version number instead of the launcher's automatically generated number

## Version 0.2.7 - (ALPHA)
### 21/08/2021 - 16:33 (GMT)
- Fixed a bug that led to the config file not being created on fresh installs or when it's deleted.

## Version 0.2.6 - (ALPHA)
### 19/08/2021 - 21:51 (GMT)
- Added the ability to launch a game when a mod is installed.
- Fixed a bug that broke mod updates.
- Style tweaks.

## Version 0.2.5 - (ALPHA)
### 16/07/2021 - 18:49 (GMT)
- Fixed mod install error relating to the progress bar.
- Updated error messages for the server page to include the Status twitter handle.
- Fixed more detailed errors not properly being displayed from mod installation.

## Version 0.2.4 - (ALPHA)
### 14/07/2021 - 21:35 (GMT)
- Added mastercomfig support to the launcher (thanks BernardoGomesNegri!)
- Servers that are down for more than 1h will now be hidden from the server list
- Added new method for ZIP extraction
- Added SCSS support
- A red button is now displayed when the launcher's auto-update information can't be received due to either internet or GitHub problems.

## Version 0.2.3 - (ALPHA)
### 26/04/2021 - 20:45 (GMT)
- Linux users are now able to get auto updates for the launcher by downloading the AppImage format, removing the need to manually download new versions.
- Fixed wrong version showing for mod updates.
- Added method to avoid the rapid navigation between mods.

## Version 0.2.2 - (ALPHA)
### 25/04/2021 - 22:35 (GMT)
- Clearer error messages for when server list is unable to operate due to website issues.
- Fixed issues with error handling for downloading and installing files.
- Fixed a mod being set as installed even if cancelled.
- The mod version box will no longer appear in the page of mods that are not installed.
- Overall UI changes to allow the launcher to dynamically resize itself.
- Password protected servers now have their data hidden.
- Internal code base cleanup. More classes changed to TS.
- Added missing server region names.
- Internal change to allow adding of new mods easier.

## Version 0.2.1 - (ALPHA)
### 28/12/2020 - 16:25 (GMT)
- Added version display to Settings page
- Added "Open Log Folder" button to settings page to help with finding logs on linux.
- Updated server mod list to be more flexable.
- Server list for Creators.TF content now includes BalanceMod servers.
- Events servers can now be seen on the server list when viewed from the "Creators.TF Events Servers" mod.
- Updated events content logo.
- Fixed bug preventing mod installation using new mod list data.
- Changed Events server content install location to `/download/` (existing installations are unaffected).
- Added cache work around to ensure launcher can get latest mod information sooner.


## Version 0.2 - (ALPHA)
### 29/11/2020 - 21:00 (GMT)
- Added Creators.TF Events Servers content to launcher. (Thanks Kaya for the artwork!)
- Added Ultimate TF2 fixes to the launcher. (Thanks agrastiOs!)
- Better error handling for Cloudflare issues
- Fixed bug with mod removal failing.
- Added support to cancel downloads and installs.
- Should have fixed error of download never finishing. (Please open an Issue if you get this still).
- Main window content is better centered.
- Fixed issue with failing to write zip file contents
- Fixed file name retrieval from downloaded files.
- Fixed some main page styling issues.


## Version 0.1.32 - (ALPHA)
### 25/10/2020 - 14:35 (GMT)
- Updated Electron, @mdi/font and marked to its latest versions (v10.1.5, v5.7.55 and v1.2.0 respectively).
- Fixed the button to update the launcher to the newest version not showing in some cases (Hopefully this time it's fixed for real).
- Mods now display their version on the top left corner of their pages.
- Applied dynamic resizing to all windows.
- Small style changes.
- Settings page:
 - Tweaked page a bit.
 - Added "Copy to Clipboard" button to easier get the config contents.


## Version 0.1.31 - (ALPHA)
### 24/09/2020 - 20:49 (GMT)
- Steam path can now be correctly located again on Linux.
- Added handle and error dialog to better explain JSON parse error.
- Added ability to re auto locate TF2 directory if a Steam path is given. (Thanks eisbaer66!).


## Version 0.1.30 - (ALPHA)
### 08/09/2020 - 22:20 (GMT)
- Updated Electron to v10.1.1.
- Improved auto locating for the Steam folder (thanks brucelay!).
- Changed server page so it can be refreshed and loaded correctly without needing to be reopened.
- Changed server page's style a bit.


## Version 0.1.20 - (ALPHA)
### 30/08/2020 - 20:04 (GMT)
- This version only applies to Windows installer users. Linux and unpacked ZIP users with v0.1.2 should not worry about updating to this version.
 - Fixed a bug where auto-update users didn't get the launcher's newest version.


## Version 0.1.2 - (ALPHA)
### 30/08/2020 - 18:16 (GMT)
- Added G-Man Announcer pack.
- Launcher's main window now dynamically adjusts itself in relation to user's monitor size on launch.

renderer.ts
const content = document.getElementById("content");
const contentDummy = document.getElementById("content-dummy");
const titleImage = document.getElementById("title-image") as HTMLImageElement;
const text = document.getElementById("content-text");
const modVersion = document.getElementById("mod-version");
const modVersionText = document.getElementById("mod-version-text");
const configPresetText = document.getElementById("config-preset-text");

const updateButton_Download = document.getElementById("update-button-download") as HTMLButtonElement;
const updateButton_Downloading = document.getElementById("update-button-downloading") as HTMLButtonElement;
const updateButton_Update = document.getElementById("update-button-update") as HTMLButtonElement;
const updateButton_Updated = document.getElementById("update-button-updated") as HTMLButtonElement;
const updateButton_Fail = document.getElementById("update-button-fail") as HTMLButtonElement;

const website = document.getElementById("socialMediaWebsite") as HTMLButtonElement;
const github = document.getElementById("socialMediaGithub") as HTMLButtonElement;
const twitter = document.getElementById("socialMediaTwitter") as HTMLButtonElement;
const discord = document.getElementById("socialMediaDiscord") as HTMLButtonElement;
const instagram = document.getElementById("socialMediaInstagram") as HTMLButtonElement;
const titleHeader = document.getElementById("title-header");
const collectionMenu = document.getElementById("collection-menu");
const collectionSelect = document.getElementById("collection-versions") as HTMLButtonElement;

const installButton = document.getElementById("install-play-button") as HTMLButtonElement;
const removeButton = document.getElementById("remove-mod") as HTMLButtonElement;
const settingsButton = document.getElementById("settings-button") as HTMLButtonElement;
const patchnotesButton = document.getElementById("patchnotes-button") as HTMLButtonElement;
const serverListButton = document.getElementById("server-list") as HTMLButtonElement;

let hasClickedInstallButton = false;

website.onclick = () => { window.ipcRenderer.send("Visit-Mod-Social", "website"); };
github.onclick = () => { window.ipcRenderer.send("Visit-Mod-Social", "github"); };
twitter.onclick = () => { window.ipcRenderer.send("Visit-Mod-Social", "twitter"); };
instagram.onclick = () => { window.ipcRenderer.send("Visit-Mod-Social", "instagram"); };
discord.onclick = () => { window.ipcRenderer.send("Visit-Mod-Social", "discord"); };

document.onload = () => {
    installButton.disabled = true;
};

const defaultBackgroundImage = "images/backgrounds/servers.jpg";

function OnClick_Mod(data) {
    if (hasClickedInstallButton) {
        return;
    }

    window.log.info("Mod entry clicked: " + data.name);

    let bgImg: string;

    if (data.backgroundimage != "") {
        bgImg = data.backgroundimage;
    } else {
        bgImg = defaultBackgroundImage;
    }

    if (bgImg.startsWith("https")) {
        content.style.backgroundImage = `url("${bgImg}")`;
    } else {
        content.style.backgroundImage = `url("${"./" + bgImg}")`;
    }

    if (data.titleimage == "") {
        titleHeader.innerText = data.name;
        titleHeader.style.display = "block";
        titleImage.style.display = "none";
    } else {
        titleImage.src = data.titleimage;
        titleImage.style.display = "block";
        titleHeader.style.display = "none";
    }

    text.innerText = data.contenttext;
    content.style.borderColor = data.bordercolor;
    content.style.backgroundPositionX = data.backgroundposX;
    content.style.backgroundPositionY = data.backgroundposY;
    content.style.backgroundBlendMode = data.backgroundBlendMode;

    contentDummy.remove();
    content.style.display = "flex";

    installButton.style.background = "";
    installButton.style.backgroundColor = "grey";
    installButton.style.color = "#EEE";
    installButton.innerText = "LOADING...";
    installButton.disabled = true;

    website.style.display = data.website != "" ? "block" : "none";
    github.style.display = data.github != "" ? "block" : "none";
    twitter.style.display = data.twitter != "" ? "block" : "none";
    instagram.style.display = data.instagram != "" ? "block" : "none";
    discord.style.display = data.discord != "" ? "block" : "none";
    serverListButton.style.display = data.serverlistproviders != "" ? "block" : "none";

    //Get the current state of this mod to set the name of the button correctly.
    //To do that, we tell the main process to set the current mod and set that up.
    window.ipcRenderer.send("SetCurrentMod", data.name);
    window.ipcRenderer.send("GetCurrentModVersion", "");
    collectionSelect.disabled = true;
    
    if (data.install.type == "githubcollection" || data.install.type == "jsonlistcollection") {
        //Do stuff for collections
        if (data.items != "") {
            collectionMenu.style.display = "block";
        } else {
            collectionMenu.style.display = "none";
        }

        //Clear the select
        collectionSelect.innerHTML = "";
        //Populate the select
        data.items.forEach((element) => {
            const opt = document.createElement("option");
            opt.value = element.itemname;
            opt.innerHTML = element.displayname;
            collectionSelect.appendChild(opt);
            if (element.default == true) {
                opt.selected = true;
            }
        });
    } else {
        collectionMenu.style.display = "none";
    }
    hasClickedInstallButton = true;
}

updateButton_Download.onclick = (downloadUpdate) => {
    window.ipcRenderer.send("download_update");
    window.log.info("User chose to download the update. Downloading it." + downloadUpdate);
};

updateButton_Update.onclick = (closeProgramAndUpdate) => {
    window.ipcRenderer.send("restart_app");
    window.log.info("User chose to restart the launcher to update." + closeProgramAndUpdate);
};

window.ipcRenderer.on("update_not_available", () => {
    window.ipcRenderer.removeAllListeners("update_not_available");
    updateButton_Updated.classList.remove("hidden");
    window.log.info("No update available... Sad!");
});

window.ipcRenderer.on("update_available", () => {
    window.ipcRenderer.removeAllListeners("update_available");
    updateButton_Updated.remove();
    updateButton_Download.classList.remove("hidden");
    window.log.info("An update is available. Waiting for user's input to actually download it.");
});

window.ipcRenderer.on("update_downloading", () => {
    window.ipcRenderer.removeAllListeners("update_downloading");
    updateButton_Download.remove();
    updateButton_Downloading.classList.remove("hidden");
    window.log.info("Downloading update...");
});

window.ipcRenderer.on("update_downloaded", () => {
    window.ipcRenderer.removeAllListeners("update_downloaded");
    updateButton_Downloading.remove();
    updateButton_Update.classList.remove("hidden");
    window.log.info("The update was downloaded and will be installed on restart. Waiting for user's input.");
});

window.ipcRenderer.on("update_error", () => {
    window.ipcRenderer.removeAllListeners("update_error");
    updateButton_Fail.classList.remove("hidden");
    window.log.info("An error occurred while trying to get update info");
});

settingsButton.addEventListener("click", () => {
    window.ipcRenderer.send("SettingsWindow", "");
});
patchnotesButton.addEventListener("click", () => {
    window.ipcRenderer.send("PatchNotesWindow", "");
});
serverListButton.addEventListener("click", () => {
    window.ipcRenderer.send("ServerListWindow", "");
});

installButton.addEventListener("click", () => {
    //Do NOT use e
    installButton.innerText = "STARTING...";
    installButton.disabled = true;
    window.ipcRenderer.send("install-play-click", collectionSelect.value);
    window.ipcRenderer.send("Open-External-Game", "gameId");
});

// Disabling stuff based on if the mod is installed or not
window.ipcRenderer.on("GetCurrentModVersion-Reply", (event, arg) => {
    if (arg != "" && arg != null) {
        modVersion.style.display = "block";
        modVersionText.innerText = "Mod version: " + arg.versionDisplay;
        configPresetText.innerText = "Config preset: " + arg.collectionversion;
        removeButton.style.display = "block";
    } else {
        modVersion.style.display = "none";
        modVersionText.innerText = "";
        configPresetText.innerText = "Config preset:";
        removeButton.style.display = "none";
    }
});

window.ipcRenderer.on("InstallButtonName-Reply", (event, arg) => {
    hasClickedInstallButton = false;
    arg = arg.toLowerCase();
    installButton.innerText = arg.toUpperCase();
    if (arg != "internal error") {
        installButton.disabled = false;
    }

    switch (arg) {
        case "installed":
            // Green (light-to-dark)
            installButton.style.background = "linear-gradient(to right, #009028 35%, #006419 75%)";
            collectionSelect.disabled = true;
            collectionSelect.style.display = "none";
            installButton.innerHTML = "<i class='mdi mdi-play'></i>PLAY";
            break;
        case "install":
            installButton.style.background = "#FF850A";
            collectionSelect.disabled = false;
            collectionSelect.style.display = "inline-flex";
            break;
        case "update":
            // Blue (dark-to-light)
            installButton.style.background = "linear-gradient(to left, #1A96FF 35%, #1A70FF 75%)";
            collectionSelect.disabled = true;
            collectionSelect.style.display = "none";
            break;
        case "internal error":
            // Red (light-to-dark)
            installButton.style.background = "linear-gradient(to right, #C72D1A 25%, #9B1100 75%)";
            collectionSelect.disabled = true;
            collectionSelect.style.display = "none";
            break;
        default:
            installButton.style.background = "grey";
            collectionSelect.disabled = true;
            collectionSelect.style.display = "none";
            break;
    }
});

window.ipcRenderer.on("FakeClickMod", (event, moddata) => {
    OnClick_Mod(moddata);
});

removeButton.addEventListener("click", (e) => {
    window.ipcRenderer.send("Remove-Mod", "");
});
